libtenzir/src/series_builder.cpp

```diff
@@ -0,0 +1,1373 @@
+///   additional logic over the inner builder type.
+///
+/// - `builder_base` is the polymorphic base class for every inner series
+///   builder for a static type kind.
+///
+/// - `typed_builder<Type>` contains the implementation for a concrete type. In
+///   particular, this class has a partial specialization for atomic types, and
+///   specializations for the list and record type.
+///
+/// - `conflict_builder` is a special builder that is used for type conflicts
+///   that could not be resolved by flushing previous events. It accepts
+///   arbitrary data and converts everything to a string.
+///
+/// - `dynamic_builder` is a wrapper over `std::unique_ptr<builder_base>` with
+///   the ability to change its type kind throughout its lifetime by replacing
+///   the inner builder with a different one.
+///
+///
+/// The following diagram summarizes the relation between all components.
+///
+///                               series_builder
+///                                     |
+///                                  (wraps)
+///                                     |
+///                            series_builder_impl
+///                               |           ^
+///                            (wraps)    (conflict)
+///                               |           ^
+///                              dynamic_builder < (contains)
+///                                     |              ^
+///                                  (stores)          ^
+///                                     |              ^
+///                                builder_base        ^
+///                               /     |      \       ^
+///                            (is)    (is)    (is)    ^            
+///                            /        |        \     ^
+///      typed_builder<int64_type>     ...      typed_builder<record_type>
+///
+///
+/// The `dynamic_builder` upgrades the typed builder when a conflict arises.
+/// To this end, we use the following lattice.
+///
+///                           confict_builder
+///                            /     |     \.
+///     typed_builder<int64_type>   ...    typed_builder<string_type>
+///                            \     |     /
+///                       typed_builder<null_type>
+///
+///
+/// Because we try to avoid a transition to `conflict_builder`, we finish the
+/// series builder for the previous elements when a conflict arises. As long as
+/// the type conflict is not caused by a list inside the current top-level
+/// series element, this resolves it. In the rare case where finishing does not
+/// suffice, we perform the upgrade to a `conflict_builder`.
+
+namespace tenzir {
+
+namespace {
+
+void check(const arrow::Status& status) {
+  TENZIR_ASSERT_CHEAP(status.ok(), status.ToString().c_str());
+}
+
+} // namespace
+
+namespace detail {
+
+class builder_base {
+public:
+  virtual ~builder_base() = default;
+
+  virtual auto finish_and_leave(int64_t count) -> typed_array = 0;
```

Comment:
```suggestion
  virtual auto finish_and_leave(int64_t num_left) -> typed_array = 0;
```

Reply:
I approached this from another direction instead: 7fe4d0f3cdb9da897134ce4e2974368ae7b45587 renames `finish_and_leave` to `finish` and changes the count from a leave-count to a finish-count. If you agree that this is clear enough, feel free to resolve.

Reply:
Yes, I think this is better. Thanks!